src/signatures/bls/mod.rs

```diff
@@ -0,0 +1,529 @@
+/// bits).
+///
+/// # Arguments
+///
+/// * `octets` - A slice of bytes representing the octet string.
+///
+/// # Returns
+///
+/// * `Ok(Usize)` corresponding to the nonnegative integer value of `octets`.
+/// * `Err(String)` if the octet string represents a number that does not fit in 256 bits.
+///
+/// # Example
+///
+/// ```
+
+/// ```
+pub fn os2ip(octets: &[u8]) -> Result<usize, String> {
+  let mut ret = 0usize;
+  for &byte in octets {
+    ret <<= 8;
+    ret += byte as usize;
+  }
+  Ok(ret)
+}
+
+// HKDF
+
+/// HKDF-Extract according to RFC 5869.
+/// If no salt is provided (i.e., salt is empty), a zero-filled salt of 32-bytes (SHA-256 output
+/// length) is used.
+pub fn hkdf_extract(salt: &[u8], ikm: &[u8]) -> Vec<u8> {
+  let salt = if salt.is_empty() {
+    // For SHA-256, the hash length is 32 bytes.
+    vec![0u8; 32]
+  } else {
+    salt.to_vec()
+  };
+  hmac_sha256(&salt, ikm).to_vec()
+}
+
+/// Implements expand_message_xmd as specified in the standard
+fn expand_message_xmd(msg: &[u8], dst: &[u8], len_in_bytes: usize) -> Vec<u8> {
+  // Parameters for SHA-256
+  const B_IN_BYTES: usize = 32; // hash digest size
+  const R_IN_BYTES: usize = 64; // hash block size
+
+  let ell = (len_in_bytes + B_IN_BYTES - 1) / B_IN_BYTES;
+  assert!(ell <= 255 && len_in_bytes <= 65535 && dst.len() <= 255);
+
+  // DST_prime = DST || I2OSP(len(DST), 1)
+  let dst_prime = [dst, &[dst.len() as u8]].concat();
+
+  // Z_pad = I2OSP(0, r_in_bytes)
+  let z_pad = vec![0u8; R_IN_BYTES];
+
+  // l_i_b_str = I2OSP(len_in_bytes, 2)
+  let l_i_b_str = i2osp(len_in_bytes, 2).unwrap();
+
+  // msg_prime = Z_pad || msg || l_i_b_str || I2OSP(0, 1) || DST_prime
+  let mut msg_prime = Vec::new();
+  msg_prime.extend_from_slice(&z_pad);
+  msg_prime.extend_from_slice(msg);
+  msg_prime.extend_from_slice(&l_i_b_str);
+  msg_prime.push(0u8);
+  msg_prime.extend_from_slice(&dst_prime);
+
+  // b_0 = H(msg_prime)
+  let mut hasher = Sha256::new();
+  hasher.update(&msg_prime);
+  let b_0 = hasher.finalize();
+
+  // b_1 = H(b_0 || I2OSP(1, 1) || DST_prime)
+  let mut hasher = Sha256::new();
+  hasher.update(&b_0);
+  hasher.update(&i2osp(1, 1).unwrap());
+  hasher.update(&dst_prime);
+  let b_1 = hasher.finalize();
+
+  let mut uniform_bytes = b_1.to_vec();
+
+  // Rest of b_vals: H(strxor(b_0, b_(i-1)) || I2OSP(i + 1, 1) || DST_prime)
+  for i in 2..=ell {
+    let mut hasher = Sha256::new();
+    let prev_b = &uniform_bytes[(i - 2) * B_IN_BYTES..(i - 1) * B_IN_BYTES];
+    let xored: Vec<u8> = b_0.iter().zip(prev_b).map(|(a, b)| a ^ b).collect();
+    hasher.update(&xored);
+    hasher.update(&i2osp(i, 1).unwrap());
+    hasher.update(&dst_prime);
+    uniform_bytes.extend_from_slice(&hasher.finalize());
+  }
+
+  uniform_bytes.truncate(len_in_bytes);
+  uniform_bytes
+}
+
+/// Implements hash_to_field as specified in the standard
+fn hash_to_field<C: EllipticCurve, D: EllipticCurve>(
```

Comment:
This also could be a nice `pub` function that could be used elsewhere. Either way, can you also add more docs here?